Spoken dialogue systems demo recipe

siddhu001 opened the pull request:
## What?

Toolkit for building demonstration of spoken dialogue systems

Modular design:

1. Wrappers for major ASR, LLM and TTS models as well as helper functions to build wrappers for open source E2E spoken dialogue systems using their publicly available checkpoints and inference codebase are detailed in `espnet2/sds` module
2. Functionality for computing major evaluation metrics at the sub-task level (e.g., ASR, text response generation, and TTS) are also provided in `espnet2/sds` module
3. `egs2/TEMPLATE/sds1` illustrates how these modules can be integrated to create a Gradio-based spoken dialogue demo.
4. Users can construct `recipes` based on this template to test specific combinations of ASR, LLM, and TTS models as building blocks for cascaded systems. These recipes also allow on-the-fly comparison using automated metrics. An example recipe is provided in `egs2/voice_assistant_demo/sds1` which is hosted at https://huggingface.co/spaces/Siddhant/Voice_Assistant_Demo.

siddhu001 commented:
Hi Shinji,

I hope you're doing well!

I’ve prepared the first draft of the PR for ESPnet-SDS and have outlined the modular design in my earlier comments. As we discussed, the current code design differs from the structure used for other tasks in the codebase. Given this, I believe the codebase may need some refactoring. I would greatly appreciate your guidance on how best to approach this process.

Thank you so much for your support and help with this!

sw005320 commented:
I have two comments.

One is whether we can stick to our normal `task` setup similar to `asr` `tts` `spk` etc, and we create a task for `sds` even if we do not include training parts. 
```
sds/{__init__.py,espnet_model.py}
```
I'm not sure that it is possible.
But if so, this would be more consistent with the others.

The second consideration is the place for the evaluation. We did not usually put it under the task, but if necessary, we can do it if the evaluation function is used for training. Can you consider the other places?

siddhu001 commented:
Hi Shinji,

Thanks a lot for the valuable feedback! Here’s how I plan to address your comments:

> `One is whether we can stick to our normal task setup similar to asr tts spk etc, and we create a task for sds even if we do not include training parts. sds/{__init__.py,espnet_model.py}`

I completely agree that maintaining a consistent task setup would improve uniformity across the codebase. I will create `sds/espnet_model.py`, which will serve as a wrapper class to integrate various components like ASR, TTS, LM, VAD, and E2E spoken dialogue modules into a structured input-output interface for spoken dialogue systems. To demonstrate this, `egs2/TEMPLATE/sds1/app.py` will create an instance of this class and combine it with evaluation metrics to present an example Gradio-based spoken dialogue demo.

> `The second consideration is the place for the evaluation. We did not usually put it under the task'

I agree! To align with the structure of other tasks, I will relocate the evaluation functions to `pyscripts/utils/`.

I will update the PR based on these changes and will notify you as soon as the revised version is ready for review. Thanks again for helping to refine this PR!

siddhu001 commented:
Hi @sw005320,

I wanted to let you know that I have updated the PR based on the changes I mentioned in my previous message to address your comments. It seems that the CIs are currently failing due to issues unrelated to this PR. I’ll wait for those to be resolved.

Please let me know if the updates look good to you or if you have have any other suggestions.

Thank you for your time and feedback!

siddhu001 commented:
Update: The PR has successfully passed all checks! 🎉

Please let me know if you have any feedback or suggestions before it’s ready to be merged.

Thank you for all your feedback!

sw005320 commented:
Thanks!
Can we come up with a better name for `voice_assistant_demo`?
The name is a bit too general.

sw005320 commented:
@Fhrozen, can you review this PR?
This is a little bit of a special task since it does not have a training component, but I think this would be a promising direction for our development as we'd have such use cases a lot in the future.

@Masao-Someki, can you also review this PR especially for the demo creation part?
This is related to our new espnet-ez project.

Fhrozen commented on egs2/TEMPLATE/asr1/pyscripts/utils/dialog_eval/ASR_WER.py:
```diff
@@ -0,0 +1,111 @@
+from espnet2.sds.utils.utils import int2float
+
+
+def handle_espnet_ASR_WER(ASR_audio_output, ASR_transcript):
```
Documentation here please

siddhu001 replied:
Sounds good! Added!

Fhrozen commented on egs2/TEMPLATE/asr1/pyscripts/utils/dialog_eval/ASR_WER.py:
```diff
@@ -0,0 +1,111 @@
+from espnet2.sds.utils.utils import int2float
+
+
+def handle_espnet_ASR_WER(ASR_audio_output, ASR_transcript):
```
Also typing

siddhu001 replied:
Sounds good! Added!

Fhrozen commented on egs2/TEMPLATE/asr1/pyscripts/utils/dialog_eval/LLM_Metrics.py:
```diff
@@ -0,0 +1,124 @@
+from transformers import AutoModel, AutoModelForCausalLM, AutoTokenizer
+
+
+def perplexity(LLM_Output):
```
typing and documentation

siddhu001 replied:
Sounds good! Added!

Fhrozen commented on egs2/TEMPLATE/asr1/pyscripts/utils/dialog_eval/LLM_Metrics.py:
```diff
@@ -0,0 +1,124 @@
+        return perplexity.cpu().item()
+
+    # Load DialoGPT model and tokenizer
+    model_name = (
```
Turn it on a argument with default vaule

siddhu001 replied:
Sounds good! Added!

Fhrozen commented on espnet2/sds/asr/espnet_asr.py:
```diff
@@ -0,0 +1,49 @@
+            dummy_input = (
+                torch.randn(
+                    (3000),
+                    dtype=getattr(torch, "float16"),
```
default dtype should become an initialization argument, so it may will allow other than float16 (bfloat16 seems more stable) or probably int or qint

siddhu001 replied:
I’ve added dtype as an initialization argument. Currently, "bfloat16" is not supported with torch.randn, so I’ve set it to "float16" as a default. Since it is now an argument, this can easily be updated in the future.

Fhrozen commented on espnet2/sds/asr/owsm_asr.py:
```diff
@@ -0,0 +1,49 @@
+            dummy_input = (
+                torch.randn(
+                    (3000),
+                    dtype=getattr(torch, "float16"),
```
ditto

siddhu001 replied:
Same as above.

Fhrozen reviewed (commented):
Excellent work. Thank you for your support and effort. I will be nice if you could add some minimal documentation and add typings to the arguments.
Also, if the models can be employed with a different dtype rather than f16 would be nice. An initialization argument that can change the dtype would be nice. I will try to test the system running it on my terminal.

Fhrozen commented on egs2/voice_assistant_demo/sds1/run.sh:
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env bash
+# Set bash to 'debug' mode, it will exit on :
+# -e 'error', -u 'undefined variable', -o ... 'error in pipeline', -x 'print commands',
+python app.py --HF_TOKEN <HF_TOKEN> --eval_options "Latency,TTS Intelligibility,TTS Speech Quality,ASR WER,Text Dialog Metrics" --tts_options kan-bayashi/ljspeech_vits,kan-bayashi/libritts_xvector_vits,kan-bayashi/vctk_multi_spk_vits,ChatTTS --llm_options meta-llama/Llama-3.2-1B-Instruct,HuggingFaceTB/SmolLM2-1.7B-Instruct  --asr_options pyf98/owsm_ctc_v3.1_1B,espnet/owsm_ctc_v3.2_ft_1B,espnet/ow ...
```
Format in multiple lines and if possible try to add whisper-tiny, or small versions

siddhu001 replied:
I formatted it to multiple lines

Fhrozen commented on egs2/voice_assistant_demo/sds1/run.sh:
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env bash
```
.gitignore requires to add entry 
`egs2/**/**/hub`

siddhu001 replied:
Thanks for catching this! 
I updated .gitignore based on your suggestion.

Fhrozen commented on egs2/voice_assistant_demo/sds1/run.sh:
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env bash
+# Set bash to 'debug' mode, it will exit on :
+# -e 'error', -u 'undefined variable', -o ... 'error in pipeline', -x 'print commands',
+python app.py --HF_TOKEN <HF_TOKEN> --eval_options "Latency,TTS Intelligibility,TTS Speech Quality,ASR WER,Text Dialog Metrics" --tts_options kan-bayashi/ljspeech_vits,kan-bayashi/libritts_xvector_vits,kan-bayashi/vctk_multi_spk_vits,ChatTTS --llm_options meta-llama/Llama-3.2-1B-Instruct,HuggingFaceTB/SmolLM2-1.7B-Instruct  --asr_options pyf98/owsm_ctc_v3.1_1B,espnet/owsm_ctc_v3.2_ft_1B,espnet/ow ...
```
it will be better to change <HF_TOKEN> to ${HF_TOKEN}, and assert that the user has already created it.

siddhu001 replied:
Incorporated this!

Masao-Someki commented:
Hi @siddhu001, thank you for your work!!
May I ask which script is especially related to the demo creation part? I will review that part!

sw005320 commented:
Thanks!
I think the main file is `egs2/TEMPLATE/sds1/app.py`
There would be some procedure script to call this.

Fhrozen commented on egs2/TEMPLATE/sds1/app.py:
```diff
@@ -0,0 +1,797 @@
+        help="Provide the possible automatic evaluation metrics available to user.",
+    )
+    parser.add_argument(
+        "--default_asr_model",
```
this values should be updated to check if included in the `<model>_options`. If value not listed in `<model>_options`, then change to first of list.

siddhu001 replied:
Incorporated this!

Fhrozen commented on espnet2/sds/espnet_model.py:
```diff
@@ -0,0 +1,256 @@
+            "_wavlm_large_raw_en_bpe5000_sp"
+        ):
+            self.s2t = ESPnetASRModel(tag=option)
+        elif option == "whisper":
```
it will be better if you replace with something like:
```
elif "whisper" in option:
            self.s2t = WhisperASRModel(tag=option.replace("whisper-", ""))
```
to handle other whisper models.

siddhu001 replied:
Sounds great! I have incorporated your suggestion!!

Fhrozen commented:
Ok, I finish a initial test, It is rea

It is really nice, the only issue is to avoid listing models that cannot be load (OutOfMemory). Adding try and skip listing may work.
Also, some instructions may be helpful, You can get lost in the app.

![Captura de pantalla 2024-12-27 023753](https://github.com/user-attachments/assets/d86e93bc-299d-48ce-9762-a1c91238f352)

BTW, when launching, the gradio url cannot transmit audio, Using VSCode with local url works fine.

FYI, I am using the following setup to run the app (you can add at the root dir of the project to launch it)
[.devcontainer.zip](https://github.com/user-attachments/files/18254507/default.devcontainer.zip)

Some entries are required at gitignore:
```
egs2/**/**/asset
egs2/**/**/checkpoint
egs2/**/**/flagged_data_points
egs2/**/**/hub
```

BTW, TTS seems the most natural, the only issues is that it keeps changing the synth voice.

Fhrozen commented on egs2/TEMPLATE/sds1/app.py:
```diff
@@ -0,0 +1,797 @@
+                    repo_type="dataset",
+                    token=access_token,
+                )
+            chat.buffer = [
```
I am reciving an error message:
```
File "/workspace/egs2/voice_assistant_demo/sds1/app.py", line 338, in transcribe
    chat.buffer = [
NameError: name 'chat' is not defined
```
from this part.

siddhu001 replied:
Hi @Fhrozen,

Thank you so much for your detailed review!

I apologize for the oversight—it was missed because I was testing without uploading user data to the Hugging Face server. I can confirm that this is indeed a bug, and I have already pushed a PR to address and fix it.

I am currently working through your remaining comments, including adding more documentation and type information. I will notify you here once I have revised the PR based on your valuable feedback.

Thank you again for your thorough and constructive feedback. I truly appreciate your help with this!

Masao-Someki commented on egs2/TEMPLATE/sds1/app.py:
```diff
@@ -0,0 +1,797 @@
+dialogue_model = None
+
+
+def read_args():
```
Maybe `parse_args`? Is there any specific reason on naming this function to be `read_args`?

siddhu001 replied:
Thanks Masao for the feedback! I have renamed the function based on your suggestion!

Masao-Someki commented on egs2/TEMPLATE/sds1/app.py:
```diff
@@ -0,0 +1,797 @@
+        description="Run the app with HF_TOKEN as a command-line argument."
+    )
+    parser.add_argument(
+        "--HF_TOKEN", required=True, help="Provide the Hugging Face token."
```
Why do we need this? I think the `HF_TOKEN` should be defined by environment variables, and not by command-line options.

siddhu001 replied:
Makes sense! I have removed this!

Masao-Someki commented on egs2/TEMPLATE/sds1/app.py:
```diff
@@ -0,0 +1,797 @@
+        "--default_llm_model",
+        required=False,
+        default="meta-llama/Llama-3.2-1B-Instruct",
+        help="Provide the default ASR model.",
```
Fix comment
```suggestion
        help="Provide the default LLM model.",
```

siddhu001 replied:
Fixed!

Masao-Someki commented on egs2/TEMPLATE/sds1/app.py:
```diff
@@ -0,0 +1,797 @@
+        "--default_tts_model",
+        required=False,
+        default="kan-bayashi/ljspeech_vits",
+        help="Provide the default ASR model.",
```
```suggestion
        help="Provide the default TTS model.",
```

siddhu001 replied:
Fixed!

Masao-Someki commented on egs2/TEMPLATE/sds1/app.py:
```diff
@@ -0,0 +1,797 @@
+        "--upload_to_hub",
+        required=False,
+        default=None,
+        help="Hugging Face dataset to upload user data",
```
Could you clarify more about this flag? I couldn't understand the role of this flag...

Masao-Someki commented on egs2/TEMPLATE/sds1/app.py:
```diff
@@ -0,0 +1,797 @@
+        help="Hugging Face dataset to upload user data",
+    )
+    args = parser.parse_args()
+    access_token = args.HF_TOKEN
```
Maybe we don't need this.

siddhu001 replied:
Removed!

Masao-Someki commented on egs2/TEMPLATE/sds1/app.py:
```diff
@@ -0,0 +1,797 @@
+audio_output = None
+audio_output1 = None
+LLM_response_arr = []
+total_response_arr = []
```
I think we should define hyper parameters first in this script, then functions, and then run those lines such as `read_args()` or `api=HfApi()`

siddhu001 replied:
Sounds good! I reordered the script to first define hyperparmaters, then the function followed by the Executable Script

Masao-Someki commented on egs2/TEMPLATE/sds1/app.py:
```diff
@@ -0,0 +1,797 @@
+
+def handle_eval_selection(
+    option, TTS_audio_output, LLM_Output, ASR_audio_output, ASR_transcript
+):
```
I think it is good to add docstrings for each functions. You can easily generate it with chatGPT.

Eg:
```
def handle_eval_selection(self, option, TTS_audio_output, LLM_Output, ASR_audio_output, ASR_transcript):
    """
    Handles the evaluation of a selected metric based on user input and provided outputs.

    Parameters:
    ----------
    option : str
        The evaluation metric selected by the user. Supported options include:
        - "Latency"
        - "TTS Intelligibility"
        - "TTS Speech Quality"
        - "ASR WER"
        - "Text Dialog Metrics"
    TTS_audio_output : np.ndarray
        The audio output generated by the TTS module for evaluation.
    LLM_Output : str
        The text output generated by the LLM module for evaluation.
    ASR_audio_output : np.ndarray
        The audio input/output used for ASR evaluation.
    ASR_transcript : str
        The transcript generated by the ASR module for evaluation.

    Returns:
    -------
    str
        A string representation of the evaluation results. The specific result depends on the selected evaluation metric:
        - "Latency": Latencies of ASR, LLM, and TTS modules.
        - "TTS Intelligibility": A score indicating how intelligible the TTS audio output is.
        - "TTS Speech Quality": A score representing the speech quality of the TTS audio output.
        - "ASR WER": The Word Error Rate (WER) of the ASR output.
        - "Text Dialog Metrics": A combination of perplexity, diversity metrics, and relevance scores for the dialog.

    Raises:
    ------
    ValueError
        If the `option` parameter does not match any supported evaluation metric.

    Example:
    -------
    >>> result = handle_eval_selection(
            option="ASR WER",
            TTS_audio_output=audio_array,
            LLM_Output="Generated response",
            ASR_audio_output=audio_input,
            ASR_transcript="Expected transcript"
        )
    >>> print(result)
    "WER: 12.34%"
    """
```

siddhu001 replied:
Sounds good! I added docstring to each function!

Masao-Someki commented on egs2/TEMPLATE/sds1/app.py:
```diff
@@ -0,0 +1,797 @@
+start_warmup()
+callback = gr.CSVLogger()
+start_record_time = None
+enable_btn = gr.Button(interactive=True, visible=True)
```
Please make a clear order in script. You can run these lines after completing defining all functions.

siddhu001 replied:
Addressed as mentioned in a comment above

Masao-Someki reviewed (commented):
@siddhu001 
Hi I have reviewed the `app.py`.

Masao-Someki commented:
I didn't check in detail, but you can mention in somewhere that this demo might not work with FireFox browser. (chrome should be fine.)

siddhu001 commented:
> Thanks! Can we come up with a better name for `voice_assistant_demo`? The name is a bit too general.

I agree that the current recipe name wasn’t ideal, so I’ve updated it to spoken_chatbot_arena for now. This name reflects its essence as a refinement of the Chatbot Arena (which is very popular in LLM space!) , where users can interact with various dialogue systems using speech and receive synthesized speech responses. If you have suggestions for a better name, please let me know!

siddhu001 commented:
> BTW, when launching, the gradio url cannot transmit audio, Using VSCode with local url works fine.

I’ve tested both the Gradio URL and the local URL on my end, and they seem to work fine, though I’ve noticed the local URL usually provides a more stable experience. Unfortunately, I wasn’t able to reproduce the issue you mentioned. It’s possible that the problem could be related to your internet connection.

I also reviewed your Docker container setup, and everything appears to be in order. To ensure compatibility, I’d recommend cross-checking the package versions in your environment with those listed in my requirements file: https://huggingface.co/spaces/Siddhant/Voice_Assistant_Demo/blob/main/requirements.txt.

siddhu001 commented:
> The only issue is to avoid listing models that cannot be load (OutOfMemory). Adding try and skip listing may work. 

Thanks for the great suggestion! I have added try, except statements.

siddhu001 commented:
> Also, some instructions may be helpful, You can get lost in the app.

Thank you for the excellent suggestion! I've added a detailed user guide to the README (https://github.com/siddhu001/espnet/blob/sds_demo_recipe/egs2/TEMPLATE/sds1/README.md) and included a link to it directly within the app for easy access. I hope this helps users navigate the app more effectively! Let me know if there are any additional improvements you’d recommend.

siddhu001 commented:
> I didn't check in detail, but you can mention in somewhere that this demo might not work with FireFox browser. (chrome should be fine.)

I added this information in the README (https://github.com/siddhu001/espnet/blob/sds_demo_recipe/egs2/TEMPLATE/sds1/README.md)

siddhu001 commented:
Hi @sw005320, @Masao-Someki, @Fhrozen,

Thank you so much for taking the time to provide such detailed feedback on this PR. I believe I’ve addressed all the suggestions, and I’m happy to share that all the CI checks have passed successfully.

Please let me know if you have any other comments before the PR’s ready to be merged. Thank you for all your feedback!

Masao-Someki reviewed (commented):
Thank you, @siddhu001!
I’ve reviewed the `app.py`, and it looks great to me!

sw005320 commented on espnet2/sds/end_to_end/mini_omni/inference.py:
Does this source come from Mini Omni?
It's better to add a reference and license in the header, as people may ignore the README file in the parent directory.

siddhu001 replied:
Yes, the source come from Mini Omni. I have added reference and license in the header!

sw005320 commented:
The PR looks good to me.
Is it possible to add some tests? (the tool requires the installation of many libraries. So, it would be difficult, but please try)

siddhu001 commented:
> The PR looks good to me. Is it possible to add some tests? (the tool requires the installation of many libraries. So, it would be difficult, but please try)

Thanks for pointing this out!  I’m currently working on adding unit tests and I’ll update the PR once the tests are ready!

sw005320 commented:
Thanks!
This PR is a great step in enhancing our inference with SDS applications.

siddhu001 commented:
Thanks a lot @sw005320 for merging this PR! I agree, I look forward to building further on this work!
